File: paleo_workbench/ui/pages/workarea_map_widget.py
"""WorkAreaMapWidget — 可复用的 GIS 工区地图（井位 + 地震工区 + 边界）。

整个面板就是一张工区图：渲染走与首页/编图一致的
:func:`~paleo_workbench.ui.qgis_stack.display_canvas.create_display_canvas`
（桥可用为 QgsMapCanvas），数据来自纯生产者
:mod:`~paleo_workbench.mapping.workarea_map_snapshot`。只读交互：滚轮缩放、
中键/空格拖拽平移、左键拾取井位（选中 + 高亮 + 激活）。

工作站的井震联合「平面图」窗格与首页共用这一套，不再维护手绘地图副本。
"""

from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import QVBoxLayout, QWidget

from paleo_workbench.mapping.workarea_map_snapshot import (
    WELLS_FLAGGED_LAYER_ID,
    WELLS_LAYER_ID,
    WORKAREA_LEGEND_ITEMS,
    build_workarea_map_snapshot,
    domain_signature,
    workarea_view_extent,
)
from paleo_workbench.ui.qgis_stack.display_canvas import create_display_canvas

# 井符号拾取容差（屏幕像素），与首页一致。
_WELL_PICK_RADIUS_PX = 16.0


class WorkAreaMapWidget(QWidget):
    """整张工区地图：set_project 喂数据，点击井位发信号。"""

    well_selected = Signal(str)    # 单击选中（well_id）
    well_activated = Signal(str)   # 单击激活（well_id）——联合窗格直接开井

    def __init__(self, parent=None, *, title: str = "", show_legend: bool = True) -> None:
        super().__init__(parent)
        self.setObjectName("WorkAreaMapWidget")
        self._project = None
        self._snapshot = None
        self._signature: object = None
        self._selected_well_id = ""
        self._title = title
        # 小窗格（如井震联合平面图）可关掉图例，避免遮挡井位。
        self._show_legend = bool(show_legend)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        self.map_canvas = create_display_canvas(parent=self)
        self.map_canvas.set_overlay_provider(self._overlay_state)
        self.map_canvas.map_clicked.connect(self._on_map_clicked)
        layout.addWidget(self.map_canvas, 1)

    def shutdown(self) -> None:
        """Stop the render backend before the host destroys the widget.

        Canvas ``shutdown`` is idempotent; hidden panes (linked
        workspace 平面图) never get a reliable ``closeEvent``, so the host
        must call this on project switch / teardown.
        """
        self.map_canvas.shutdown()

    # ------------------------------------------------------------------
    # data
    # ------------------------------------------------------------------

    def set_project(self, project) -> None:
        """Rebind the project; snapshot rebuilds only on domain change."""
        self._project = project
        if project is None:
            self._snapshot = None
            self._signature = None
            return
        signature = domain_signature(project)
        if signature == self._signature:
            return
        self._signature = signature
        snapshot = build_workarea_map_snapshot(project)
        self._snapshot = snapshot
        self.map_canvas.set_layer_snapshot(snapshot)
        extent = workarea_view_extent(snapshot)
        if extent is not None:
            self.map_canvas.set_extent(extent)

    def zoom_to_all(self) -> None:
        if self._snapshot is None:
            return
        extent = workarea_view_extent(self._snapshot)
        if extent is not None:
            self.map_canvas.set_extent(extent)

    # ------------------------------------------------------------------
    # selection
    # ------------------------------------------------------------------

    def select_well(self, well_id: str, *, zoom: bool = False, emit: bool = False) -> None:
        """Highlight a well (yellow ring via overlay); optionally zoom to it."""
        well_id = str(well_id or "")
        self._selected_well_id = well_id
        feature = self._well_feature(well_id)
        if zoom and feature is not None:
            coords = (feature.get("geometry") or {}).get("coordinates") or ()
            if len(coords) >= 2:
                x, y = float(coords[0]), float(coords[1])
                half = self._current_half_span() * 0.2
                self.map_canvas.set_extent((x - half, y - half, x + half, y + half))
        self.map_canvas.update()
        if emit and well_id:
            self.well_selected.emit(well_id)

    def selected_well_id(self) -> str:
        return self._selected_well_id

    # ------------------------------------------------------------------
    # internals
    # ------------------------------------------------------------------

    def _overlay_state(self) -> dict:
        elements = ["比例尺", "指北针"]
        if self._title:
            elements.append("标题栏")
        if self._show_legend:
            elements.append("图例")
        state: dict = {
            "decorations": {
                "title": self._title,
                "elements": tuple(elements),
                "legend_items": [
                    {"label": label, "color": color} for label, color in WORKAREA_LEGEND_ITEMS
                ],
            }
        }
        feature = self._well_feature(self._selected_well_id)
        if feature is not None:
            state["selected_features"] = (feature,)
        return state

    def _well_feature(self, well_id: str) -> dict | None:
        if not well_id or self._snapshot is None:
            return None
        for layer in self._snapshot.layers:
            if layer.id not in (WELLS_LAYER_ID, WELLS_FLAGGED_LAYER_ID):
                continue
            for feature in layer.features:
                props = feature.get("properties") or {}
                if str(props.get("well_id") or "") == well_id:
                    return feature
        return None

    def _current_half_span(self) -> float:
        xmin, ymin, xmax, ymax = self.map_canvas.view_extent
        return max((xmax - xmin) / 2.0, (ymax - ymin) / 2.0, 1.0)

    def _on_map_clicked(self, point) -> None:
        """Left-click: pick the nearest well within the screen tolerance."""
        if self._snapshot is None:
            return
        try:
            click_screen = self.map_canvas.map_to_screen(
                (float(point[0]), float(point[1]))
            )
        except (TypeError, ValueError, IndexError, ArithmeticError):
            # #1166: ZeroDivisionError 是 ArithmeticError（退化 extent），
            # 原清单漏了它——“以为守了其实没守”。
            return
        best_id = ""
        best_dist = _WELL_PICK_RADIUS_PX
        for layer in self._snapshot.layers:
            if layer.id not in (WELLS_LAYER_ID, WELLS_FLAGGED_LAYER_ID):
                continue
            for feature in layer.features:
                geometry = feature.get("geometry") or {}
                coords = geometry.get("coordinates")
                if geometry.get("type") != "Point" or not coords:
                    continue
                screen = self.map_canvas.map_to_screen(
                    (float(coords[0]), float(coords[1]))
                )
                dx = screen.x() - click_screen.x()
                dy = screen.y() - click_screen.y()
                dist = (dx * dx + dy * dy) ** 0.5
                if dist <= best_dist:
                    best_dist = dist
                    best_id = str((feature.get("properties") or {}).get("well_id") or "")
        if best_id:
            self.select_well(best_id)
            self.well_selected.emit(best_id)
            self.well_activated.emit(best_id)

File: libs/ui_map/src/workarea_map_widget.cpp
#include <pwb/ui_map/workarea_map_widget.hpp>

#include <cmath>
#include <utility>

#include <QPointF>
#include <QVBoxLayout>

#include <pwb/ui_map/display_map_canvas.hpp>

namespace pwb::ui_map {

namespace {

QString qstr(const std::string& s) {
    return QString::fromUtf8(s.data(), static_cast<qsizetype>(s.size()));
}

}  // namespace

WorkAreaMapWidget::WorkAreaMapWidget(QWidget* parent, std::string title,
                                     bool show_legend)
    : QWidget(parent), title_(std::move(title)), show_legend_(show_legend) {
    setObjectName(QStringLiteral("WorkAreaMapWidget"));

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);

    map_canvas_ = new DisplayMapCanvas(this);
    map_canvas_->set_overlay_provider([this] { return overlay_state(); });
    connect(map_canvas_, &DisplayMapCanvas::map_clicked, this,
            [this](double x, double y) { on_map_clicked(x, y); });
    layout->addWidget(map_canvas_, 1);
}

void WorkAreaMapWidget::shutdown() { map_canvas_->shutdown(); }

// ---------------------------------------------------------------------------
// data
// ---------------------------------------------------------------------------

void WorkAreaMapWidget::set_project(
    const Json& project,
    const std::function<Json(const Json&)>& snapshot_builder) {
    if (project.is_null()) {
        snapshot_ = Json(nullptr);
        signature_ = Json(nullptr);
        return;
    }
    const Json signature = domain_signature(project);
    if (!signature_.is_null() && signature == signature_) {
        return;
    }
    signature_ = signature;
    Json snapshot =
        snapshot_builder ? snapshot_builder(project) : Json(nullptr);
    snapshot_ = snapshot;
    map_canvas_->set_layer_snapshot(snapshot);
    const std::optional<Extent> extent = workarea_view_extent(snapshot);
    if (extent.has_value()) {
        map_canvas_->set_extent(*extent);
    }
}

void WorkAreaMapWidget::zoom_to_all() {
    if (snapshot_.is_null()) {
        return;
    }
    const std::optional<Extent> extent = workarea_view_extent(snapshot_);
    if (extent.has_value()) {
        map_canvas_->set_extent(*extent);
    }
}

// ---------------------------------------------------------------------------
// selection
// ---------------------------------------------------------------------------

void WorkAreaMapWidget::select_well(const std::string& well_id, bool zoom,
                                    bool emit_signal) {
    selected_well_id_ = well_id;
    const Json feature = well_feature(well_id);
    if (zoom && feature.is_object()) {
        const Json coords = field_value(
            field_value(feature, "geometry", Json::object()),
            "coordinates", Json::array());
        if (coords.is_array() && coords.size() >= 2) {
            try {
                const double x = coords.at(0).get<double>();
                const double y = coords.at(1).get<double>();
                const double half = current_half_span() * 0.2;
                map_canvas_->set_extent(
                    Extent{x - half, y - half, x + half, y + half});
            } catch (const Json::exception&) {
                // Non-numeric coordinates — Python float() would raise and
                // the page lets it propagate; the widget stays put.
            }
        }
    }
    map_canvas_->update();
    if (emit_signal && !well_id.empty()) {
        emit well_selected(qstr(well_id));
    }
}

// ---------------------------------------------------------------------------
// internals
// ---------------------------------------------------------------------------

Json WorkAreaMapWidget::overlay_state() const {
    return workarea_overlay_state(title_, show_legend_,
                                  well_feature(selected_well_id_));
}

Json WorkAreaMapWidget::well_feature(const std::string& well_id) const {
    if (well_id.empty() || snapshot_.is_null()) {
        return Json(nullptr);
    }
    return well_feature_by_id(snapshot_, well_id);
}

double WorkAreaMapWidget::current_half_span() const {
    const Extent extent = map_canvas_->view_extent();
    return std::max({(extent[2] - extent[0]) / 2.0,
                     (extent[3] - extent[1]) / 2.0, 1.0});
}

void WorkAreaMapWidget::on_map_clicked(double x, double y) {
    // Left-click: pick the nearest well within the screen tolerance.
    if (snapshot_.is_null()) {
        return;
    }
    const std::optional<QPointF> click_screen =
        map_canvas_->map_to_screen(x, y);
    if (!click_screen.has_value()) {
        // Python catches ArithmeticError here (#1166: a degenerate extent
        // used to leak a ZeroDivisionError through this path).
        return;
    }
    std::vector<WellPickCandidate> candidates;
    for (const WellFeature& well : well_candidates_from_snapshot(snapshot_)) {
        const std::optional<QPointF> screen =
            map_canvas_->map_to_screen(well.x, well.y);
        if (!screen.has_value()) {
            continue;
        }
        candidates.push_back(
            WellPickCandidate{well.well_id, screen->x(), screen->y()});
    }
    const std::string best = pick_well_id(
        {click_screen->x(), click_screen->y()}, candidates,
        kWellPickRadiusPx);
    if (!best.empty()) {
        select_well(best);
        emit well_selected(qstr(best));
        emit well_activated(qstr(best));
    }
}

}  // namespace pwb::ui_map
